src/herder/test/HerderTests.cpp

```diff
@@ -1617,6 +1617,206 @@ TEST_CASE("SCP State", "[herder][acceptance]")
+        auto ct = std::max<uint64>(
+            lcl.header.scpValue.closeTime + 1,
+            VirtualClock::to_time_t(validator->getClock().now()));
+        StellarValue sv(txSet->getContentsHash(), ct, emptyUpgradeSteps,
+                        STELLAR_VALUE_BASIC);
+        SCPEnvelope envelope;
+
+        auto& st = envelope.statement;
+        st.slotIndex = lcl.header.ledgerSeq + 1;
+        st.pledges.type(SCP_ST_EXTERNALIZE);
+        auto& ext = envelope.statement.pledges.externalize();
+        ext.commit.counter = UINT32_MAX;
+        ext.commit.value = xdr::xdr_to_opaque(sv);
+        ext.nH = UINT32_MAX;
+        ext.commitQuorumSetHash = qSetHash;
+
+        st.nodeID = validatorKey.getPublicKey();
+        envelope.signature = validatorKey.sign(xdr::xdr_to_opaque(
+            validator->getNetworkID(), ENVELOPE_TYPE_SCP, st));
+
+        LedgerCloseData lcd{lcl.header.ledgerSeq + 1, txSet, sv};
+        validator->getLedgerManager().closeLedger(lcd);
+
+        validatorSCPMessages.emplace_back(std::make_pair(envelope, txSet));
+    }
+
+    HerderImpl& herder = *static_cast<HerderImpl*>(&listener->getHerder());
+    auto const& lm = listener->getLedgerManager();
+    auto const& cm = listener->getCatchupManager();
+
+    REQUIRE(herder.getCurrentLedgerSeq() == currentListenerLedger());
+
+    auto testOutOfOrder = [&](bool partial) {
+        // This should trigger CatchupManager to start buffering ledgers
+        auto futureSlot = validatorSCPMessages[NUM_MESSAGES - 1];
+        REQUIRE(herder.recvSCPEnvelope(futureSlot.first, qset,
+                                       *(futureSlot.second)) ==
+                Herder::ENVELOPE_STATUS_READY);
+
+        // Wait until listener goes out of sync
+        simulation->crankUntil(
+            [&]() { return !listener->getLedgerManager().isSynced(); },
+            2 * Herder::CONSENSUS_STUCK_TIMEOUT_SECONDS, false);
+
+        // Ensure LM is out of sync, and Herder tracks ledger seq from latest
+        // envelope
+        REQUIRE(herder.getCurrentLedgerSeq() ==
+                futureSlot.first.statement.slotIndex);
+        REQUIRE(!lm.isSynced());
+
+        // Next, externalize a contiguous ledger
+        // This will cause LM to apply it, and catchup manager will try to apply
+        // buffered ledgers
+        auto contiguousSlot = validatorSCPMessages[0];
+
+        REQUIRE(herder.recvSCPEnvelope(contiguousSlot.first, qset,
+                                       *(contiguousSlot.second)) ==
+                Herder::ENVELOPE_STATUS_READY);
+
+        // Ensure Herder still tracks the right index despite receiving older
+        // message
+        REQUIRE(herder.getCurrentLedgerSeq() ==
+                futureSlot.first.statement.slotIndex);
+
+        // Ledger manager is still out of sync, as we have seen future ledgers
+        REQUIRE(!lm.isSynced());
+
+        // complete - all messages are received out of order (10, 7, 9, 8)
```

Comment:
should we just pass it the actual vector of indexes to apply instead of a bool? This is pretty hard to follow. I'm thinking you'd pass in `{10,7, 9,8}` and `{10, 7, 8, 9}`